lib/SILGen/SILGenPattern.cpp

```diff
@@ -1032,17 +1032,25 @@ void PatternMatchEmission::emitWildcardDispatch(ClauseMatrix &clauses,
   assert(!hasGuard || !clauses[row].isIrrefutable());
 
   auto stmt = clauses[row].getClientData<CaseStmt>();
```

Comment:
It sounds incorrect that the pointer is cast to `CaseStmt` here, when it in fact isn't.  I think you should be casting to `Stmt` here, and then downcasting even further below (with `dyn_cast<CaseStmt>`).

In fact, you can add implement `getClientData()` with `return cast<T>(ClientData)` to prevent such issues in future (and maybe catch other cases in our codebase).